File: Source/Aladdin.CAPI.STB/CAPI/STB/Sign/STB11762/SignData.cs
using System;

namespace Aladdin.CAPI.STB.Sign.STB11762
{
    ///////////////////////////////////////////////////////////////////////////
    // Алгоритм выработки подписи данных СТБ 1176.2
    ///////////////////////////////////////////////////////////////////////////
    public class SignData : CAPI.SignData
    {
        // способ кодирования чисел
        private const Math.Endian Endian = Math.Endian.LittleEndian; 

        // алгоритм хэширования и секретный параметр
        private CAPI.Hash hashAlgorithm; private Math.BigInteger K;

        // конструктор
        public SignData() { hashAlgorithm = null; K = null; }

        // освободить используемые ресурсы
        protected override void OnDispose()
        {
            // освободить используемые ресурсы
            RefObject.Release(hashAlgorithm); base.OnDispose(); 
        }
        // инициализировать алгоритм
        public override void Init(IPrivateKey privateKey, IRand rand) 
        {
            // освободить используемые ресурсы
            RefObject.Release(hashAlgorithm); hashAlgorithm = null; 

            // преобразовать тип ключа
            STB.STB11762.IBDSPrivateKey stbPrivateKey = 
                (STB.STB11762.IBDSPrivateKey)privateKey;

            // преобразовать тип параметров
            STB.STB11762.IBDSParameters parameters = 
                (STB.STB11762.IBDSParameters)privateKey.Parameters; 

            // прочитать параметры алгоритма
            Math.BigInteger P = parameters.P; Math.BigInteger Q = parameters.Q;
            Math.BigInteger A = parameters.G; int             R = parameters.R;
            
            // вызвать базовую функцию
            base.Init(privateKey, rand); K = Math.BigInteger.Zero; 

            // указать генератор случайных чисел
            using (Random random = new Random(rand))
            do {             
                // сгенерировать число 0 < K < Q
                K = new Math.BigInteger(R, random);
            }
            // проверить выполнение требуемых условий
            while (K.Signum == 0 || K.CompareTo(Q) >= 0); 

            // вычислить T = A^(K)
            Math.BigInteger T = (new Math.Fp.MontGroup(P)).Power(A, K);

            // закодировать число T
            byte[] encodedT = Math.Convert.FromBigInteger(T, Endian);  

            // создать алгоритм хэширования
            hashAlgorithm = CreateHashAlgorithm(privateKey, parameters.H); 

            // прохэшировать число T
            hashAlgorithm.Init(); hashAlgorithm.Update(encodedT, 0, encodedT.Length); 
        }
        // обработать данные
        public override void Update(byte[] data, int dataOff, int dataLen)
        {
            // прохэшировать данные
            hashAlgorithm.Update(data, dataOff, dataLen); 
        }
        // получить подпись данных
        public override byte[] Finish(IRand rand)
        {
            // преобразовать тип ключа
            STB.STB11762.IBDSPrivateKey stbPrivateKey = 
                (STB.STB11762.IBDSPrivateKey)PrivateKey;

            // преобразовать тип параметров
            STB.STB11762.IBDSParameters parameters = 
                (STB.STB11762.IBDSParameters)PrivateKey.Parameters; 

            // прочитать параметры алгоритма
            Math.BigInteger Q = parameters.Q; int R = parameters.R;

            // получить хэш-значение
            byte[] H = new byte[32]; hashAlgorithm.Finish(H, 0);

            // обнулить незначащие биты
            if (((R - 1) % 8) != 0) H[(R - 1) / 8] &= (byte)((1 << ((R - 1) % 8)) - 1);

            // преобразовать хэш-значение в число U
            Math.BigInteger U = Math.Convert.ToBigInteger(H, 0, (R + 6) / 8, Endian);

            // выполнить вычисления
            Math.BigInteger XU = stbPrivateKey.X.Multiply(U).Mod(Q);

            // вычислить число V
            Math.BigInteger V = (K.CompareTo(XU) > 0) ? K.Subtract(XU) : K.Add(Q).Subtract(XU);

            // освободить используемые ресурсы
            RefObject.Release(hashAlgorithm); hashAlgorithm = null;  

            // проверить корректность
            if (U.Signum == 0 || V.Signum == 0) throw new InvalidOperationException();

            // выполнить конкатенацию 
            Math.BigInteger UV = U.ShiftLeft(R).Add(V); 

            // вычислить подпись
            byte[] signature = Math.Convert.FromBigInteger(UV, Endian, (R + 3) / 4); 

            // вернуть вычисленную подпись
            base.Finish(rand); return signature; 
        }
        // создать алгоритм хэширования
        protected CAPI.Hash CreateHashAlgorithm(IPrivateKey privateKey, byte[] start) 
        { 
            // создать алгоритм хэширования
            return new Hash.STB11761(start);     
        }
    }
}

File: Source/Aladdin.CAPI.STB.Java/src/aladdin/capi/stb/sign/stb11762/SignData.java
package aladdin.capi.stb.sign.stb11762;
import aladdin.*; 
import aladdin.math.*; 
import aladdin.math.Fp.*;
import aladdin.capi.*;
import aladdin.capi.stb.stb11762.*;
import java.io.*;
import java.math.*;

///////////////////////////////////////////////////////////////////////////
// Алгоритм выработки подписи данных СТБ 1176.2
///////////////////////////////////////////////////////////////////////////
public class SignData extends aladdin.capi.SignData
{
    // способ кодирования чисел
    private static final Endian ENDIAN = Endian.LITTLE_ENDIAN; 
    
    // алгоритм хэширования и секретный параметр
    private Hash hashAlgorithm;	private BigInteger K;
    
    // конструктор
    public SignData() { hashAlgorithm = null; K = null; }
    
    // освободить используемые ресурсы
    @Override protected void onClose() throws IOException  
    {
        // освободить используемые ресурсы
        RefObject.release(hashAlgorithm); super.onClose(); 
    }
    // инициализировать алгоритм
    @Override public void init(IPrivateKey privateKey, IRand rand) throws IOException 
    { 
        // освободить используемые ресурсы
        RefObject.release(hashAlgorithm); hashAlgorithm = null; 
        
        // преобразовать тип параметров
        IBDSParameters parameters = (IBDSParameters)privateKey.parameters(); 
        
        // прочитать параметры алгоритма
        BigInteger P = parameters.bdsP(); BigInteger Q = parameters.bdsQ();
        BigInteger A = parameters.bdsA(); int        R = parameters.bdsR();
        
        // выполнить базовую функцию
        super.init(privateKey, rand); K = BigInteger.ZERO; 

        // указать генератор случайных чисел
        try (Random random = new Random(rand)) 
        {
            // до выполнения требуемых условий
            while (K.signum() == 0 || K.compareTo(Q) >= 0)
            {
                // сгенерировать число 0 < K < Q
                K = new BigInteger(R, random);
            }
        }
        // вычислить T = A^(K)
        BigInteger T = (new MontGroup(P)).power(A, K);

        // закодировать число T
        byte[] encodedT = Convert.fromBigInteger(T, ENDIAN);  

        // создать алгоритм хэширования
        hashAlgorithm = createHashAlgorithm(privateKey, parameters.bdsH()); 

        // прохэшировать число T
        hashAlgorithm.init(); hashAlgorithm.update(encodedT, 0, encodedT.length); 
    }
    // обработать данные
    @Override public void update(byte[] data, int dataOff, int dataLen) throws IOException
    {
        // прохэшировать данные
        hashAlgorithm.update(data, dataOff, dataLen); 
    }
    // получить подпись данных
    @Override public byte[] finish(IRand rand) throws IOException
    {
        // преобразовать тип ключа
        IBDSPrivateKey stbPrivateKey = (IBDSPrivateKey)privateKey(); 
        
        // преобразовать тип параметров
        IBDSParameters parameters = (IBDSParameters)stbPrivateKey.parameters(); 
        
        // прочитать параметры алгоритма
        BigInteger Q = parameters.bdsQ(); int R = parameters.bdsR();

        // получить хэш-значение
        byte[] H = new byte[32]; hashAlgorithm.finish(H, 0);
        
        // обнулить незначащие биты
        if (((R - 1) % 8) != 0) H[(R - 1) / 8] &= (byte)((1 << ((R - 1) % 8)) - 1);

        // преобразовать хэш-значение в число U
        BigInteger U = Convert.toBigInteger(H, 0, (R + 6) / 8, ENDIAN);

        // выполнить вычисления
        BigInteger XU = stbPrivateKey.bdsX().multiply(U).mod(Q);

        // вычислить число V
        BigInteger V = (K.compareTo(XU) > 0) ? K.subtract(XU) : K.add(Q).subtract(XU);

        // освободить выделенные ресурсы
        RefObject.release(hashAlgorithm); hashAlgorithm = null; 

        // проверить корректность
        if (U.signum() == 0 || V.signum() == 0) throw new IllegalStateException(); 

        // выполнить конкатенацию 
        BigInteger UV = U.shiftLeft(R).add(V); 

        // вычислить подпись
        byte[] signature = Convert.fromBigInteger(UV, ENDIAN, (R + 3) / 4); 
        
        // вернуть вычисленную подпись
        super.finish(rand); return signature; 
    }
    // создать алгоритм хэширования
    protected Hash createHashAlgorithm(IPrivateKey privateKey, byte[] start) throws IOException
    { 
        // создать алгоритм хэширования
        return new aladdin.capi.stb.hash.STB11761(start);     
    }
}
